src/entry.c

```diff
@@ -72,109 +82,123 @@ sds entryGetField(const entry *entry) {
+static inline long long *entryGetExpiryRef(const entry *entry) {
+    debugServerAssert(entryHasExpiry(entry));
+    char *buf = entryGetAllocPtr(entry);
+    debugServerAssert((((uintptr_t)buf & 0x7) == 0)); /* Test that the allocation is indeed 8 bytes aligned
+                                                       * This is needed since we access the expiry as with pointer casting
+                                                       * which require the access to be 8 bytes aligned. */
+    return (long long *)buf;
+}
 
 /* Returns the entry expiration timestamp.
  * In case this entry has no expiration time, will return EXPIRE_NONE. */
 long long entryGetExpiry(const entry *entry) {
-    long long expiry = EXPIRY_NONE;
-    if (entryHasExpiry(entry)) {
-        char *buf = entryGetAllocPtr(entry);
-        debugServerAssert((((uintptr_t)buf & 0x7) == 0)); /* Test that the allocation is indeed 8 bytes aligned
-                                                           * This is needed since we access the expiry as with pointer casting
-                                                           * which require the access to be 8 bytes aligned. */
-        expiry = *(long long *)buf;
-    }
-    return expiry;
+    if (entryHasExpiry(entry)) return *entryGetExpiryRef(entry);
+    return EXPIRY_NONE;
 }
 
 /* Modify the expiration time of this entry and return a pointer to the (potentially new) entry. */
 entry *entrySetExpiry(entry *e, long long expiry) {
-    if (entryHasExpiry(e)) {
-        char *buf = entryGetAllocPtr(e);
-        debugServerAssert((((uintptr_t)buf & 0x7) == 0)); /* Test that the allocation is indeed 8 bytes aligned
-                                                           * This is needed since we access the expiry as with pointer casting
-                                                           * which require the access to be 8 bytes aligned. */
-        *(long long *)buf = expiry;
+    if (expiry != EXPIRY_NONE && entryHasExpiry(e)) {
+        *entryGetExpiryRef(e) = expiry;
         return e;
     }
-    entry *new_entry = entryUpdate(e, NULL, expiry);
-    return new_entry;
+    if (entryHasStringRef(e)) {
+        stringRef *value = (stringRef *)*entryGetValueRef(e);
+        return entryUpdateAsStringRef(e, value->buf, value->len, expiry);
+    }
+    return entryUpdate(e, NULL, expiry);
 }
 
 /* Return true in case the entry has assigned expiration or false otherwise. */
 bool entryIsExpired(entry *entry) {
     return timestampIsExpired(entryGetExpiry(entry));
 }
 /**************************************** Entry Expiry API - End *****************************************/
+/* Frees the entry non-embedded value */
+static inline void entryFreeValuePtr(entry *entry) {
+    serverAssert(entryHasValuePtr(entry));
+    void **value_ref = entryGetValueRef(entry);
+    if (entryHasStringRef(entry)) {
+        zfree(*value_ref);
+        return;
+    }
+    sdsfree(*value_ref);
```

Comment:
Since this code is freeing the value reference and NOT freeing the entry, we should be nulling out the pointer for safety.
Eliminate the extra return (use an else).  Set `*value_ref = NULL` after performing the free.
```suggestion
    if (entryHasStringRef(entry)) {
        zfree(*value_ref);
    } else {
        sdsfree(*value_ref);
    }
    *value_ref = NULL;
```

Reply:
This seems like an issue with the existing [code](https://github.com/valkey-io/valkey/blob/unstable/src/entry.c#L431). Anyway,  I'll fix this.